test/e2e/domainmapping/domain_mapping_websocket_test.go

```diff
@@ -0,0 +1,127 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package domainmapping
+
+import (
+	"context"
+	"net/url"
+	"strings"
+	"testing"
+
+	"knative.dev/pkg/test/spoof"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	duckv1 "knative.dev/pkg/apis/duck/v1"
+	pkgTest "knative.dev/pkg/test"
+	"knative.dev/serving/pkg/apis/serving/v1alpha1"
+	"knative.dev/serving/test"
+
+	v1test "knative.dev/serving/test/v1"
+)
+
+const (
+	wsServerTestImageName = "wsserver"
+)
+
+func TestDomainMappingWebsocket(t *testing.T) {
+	if !test.ServingFlags.EnableBetaFeatures {
+		t.Skip("Beta features not enabled")
+	}
+
+	if !strings.Contains(test.ServingFlags.IngressClass, "kourier") {
+		t.Skip("Skip this test for non-kourier ingress.")
+	}
+
+	t.Parallel()
+	clients := test.Setup(t)
+
+	names := test.ResourceNames{
+		Service: test.ObjectNameForTest(t),
+		Image:   wsServerTestImageName,
+	}
+
+	test.EnsureTearDown(t, clients, &names)
+	ctx := context.Background()
+
+	ksvc, err := v1test.CreateServiceReady(t, clients, &names)
+	if err != nil {
+		t.Fatalf("Failed to create initial Service %v: %v", names.Service, err)
+	}
+
+	host := ksvc.Service.Name + ".example.org"
+	resolvableCustomDomain := false
+
+	if test.ServingFlags.CustomDomain != "" {
+		host = ksvc.Service.Name + "." + test.ServingFlags.CustomDomain
+		resolvableCustomDomain = true
+	}
+
+	dm := v1alpha1.DomainMapping{
+		TypeMeta: metav1.TypeMeta{},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        host,
+			Namespace:   ksvc.Service.Namespace,
+			Annotations: map[string]string{"kourier.knative.dev/disable-http2": "true"},
+		},
+		Spec: v1alpha1.DomainMappingSpec{
+			Ref: duckv1.KReference{
+				APIVersion: "serving.knative.dev/v1",
+				Name:       ksvc.Service.Name,
+				Namespace:  ksvc.Service.Namespace,
+				Kind:       "Service",
+			},
+		},
+		Status: v1alpha1.DomainMappingStatus{},
+	}
+
+	_, err = clients.ServingAlphaClient.DomainMappings.Create(ctx, &dm, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("Problem creating DomainMapping %q: %v", host, err)
+	}
+	t.Cleanup(func() {
+		clients.ServingAlphaClient.DomainMappings.Delete(ctx, dm.Name, metav1.DeleteOptions{})
+	})
+
+	waitErr := wait.PollImmediate(test.PollInterval, test.PollTimeout, func() (bool, error) {
+		var err error
+		dm, err := clients.ServingAlphaClient.DomainMappings.Get(ctx, dm.Name, metav1.GetOptions{})
+		if err != nil {
+			return true, err
+		}
+
+		return dm.IsReady(), nil
+	})
+	if waitErr != nil {
+		t.Fatalf("The DomainMapping %s was not marked as Ready: %v", dm.Name, waitErr)
+	}
+
+	_, err = pkgTest.CheckEndpointState(ctx,
+		clients.KubeClient,
+		t.Logf,
+		&url.URL{Scheme: "ws", Host: host},
+		spoof.MatchesBody(wsServerTestImageName),
+		"DomainMappingWithWebsocket",
+		resolvableCustomDomain,
+	)
```

Comment:
```suggestion
                test.AddRootCAtoTransport(context.Background(), t.Logf, clients, test.ServingFlags.HTTPS),
	)
```

I think we need to add RootCA to the request for HTTPS test.

Reply:
Actually I am confused about this part. 
1. I think the protocol should be "ws", or "wss" with crypto. What do you mean by https?
2. With my local test, the `CheckEndpointState` function does not support websocket, and it failed.
So do we need to support websocket in pkg project?

Reply:
please check this: https://github.com/knative/serving/runs/7556816067?check_suite_focus=true#step:11:5554

Reply:
Ah, thank you.
Yeah, you are correct. It seems we cannot use `CheckEndpointState` so we need 
 to other way to validate it. I guess we can validate it with custom func `validateWebSocketConnection` like https://github.com/knative/serving/blob/main/test/e2e/websocket_test.go#L48 ?